src/EditorFeatures/Core/ExtractMethod/ExtractMethodCommandHandler.cs

```diff
@@ -133,59 +171,67 @@ public bool ExecuteCommand(ExtractMethodCommandArgs args, CommandExecutionContex
-                    return true;
+                    return;
                 }
             }
 
-            var cleanupOptions = document.GetCodeCleanupOptionsAsync(_globalOptions, cancellationToken).AsTask().WaitAndGetResult(cancellationToken);
+            var cleanupOptions = await document.GetCodeCleanupOptionsAsync(_globalOptions, cancellationToken).ConfigureAwait(false);
+            var (formattedDocument, methodNameAtInvocation) = await result.GetFormattedDocumentAsync(cleanupOptions, cancellationToken).ConfigureAwait(false);
 
-            // apply the change to buffer
-            // get method name token
-            SyntaxToken methodNameAtInvocation;
-            using (var undoTransaction = _undoManager.GetTextBufferUndoManager(textBuffer).TextBufferUndoHistory.CreateTransaction("Extract Method"))
-            {
-                // apply extract method code to buffer
-                (var formattedDocument, methodNameAtInvocation) = result.GetFormattedDocumentAsync(cleanupOptions, cancellationToken).WaitAndGetResult(cancellationToken);
-                formattedDocument.Project.Solution.Workspace.ApplyDocumentChanges(formattedDocument, cancellationToken);
-
-                // apply changes
-                undoTransaction.Complete();
-            }
+            await ApplyChangeAsync(formattedDocument, textBuffer, waitContext).ConfigureAwait(false);
 
             // start inline rename
-            var snapshotAfterFormatting = textBuffer.CurrentSnapshot;
-            var documentAfterFormatting = snapshotAfterFormatting.GetOpenDocumentInCurrentContextWithChanges();
-            _renameService.StartInlineSession(documentAfterFormatting, methodNameAtInvocation.Span, cancellationToken);
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var textSnapshot = textBuffer.CurrentSnapshot;
+            document = textSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document != null)
+                _renameService.StartInlineSession(document, methodNameAtInvocation.Span, cancellationToken);
 
             // select invocation span
-            view.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(snapshotAfterFormatting, methodNameAtInvocation.Span.End));
-            view.SetSelection(
-                methodNameAtInvocation.Span.ToSnapshotSpan(snapshotAfterFormatting));
+            view.TryMoveCaretToAndEnsureVisible(new SnapshotPoint(textSnapshot, methodNameAtInvocation.Span.End));
+            view.SetSelection(methodNameAtInvocation.Span.ToSnapshotSpan(textSnapshot));
+        }
 
-            return true;
+        private async Task ApplyChangeAsync(
+            Document formattedDocument, ITextBuffer textBuffer, IBackgroundWorkIndicatorContext waitContext)
+        {
+            var cancellationToken = waitContext.UserCancellationToken;
+
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
it might make sense to pull this out of this method into the caller - then you just have 1 switch to UI thread before applying changes and starting inline rename

Reply:
sounds good.

Reply:
done.

Reply:
done.